core/src/main/scala/kafka/server/ClientQuotaManager.scala

```diff
@@ -53,11 +58,73 @@ object ClientQuotaManagerConfig {
   val DefaultQuotaWindowSizeSeconds = 1
   // Purge sensors after 1 hour of inactivity
   val InactiveSensorExpirationTimeSeconds  = 3600
+
+  val UnlimitedQuota = Quota.upperBound(Long.MaxValue)
+  val DefaultClientIdQuotaId = QuotaId(None, Some(ConfigEntityName.Default))
+  val DefaultUserQuotaId = QuotaId(Some(ConfigEntityName.Default), None)
+  val DefaultUserClientIdQuotaId = QuotaId(Some(ConfigEntityName.Default), Some(ConfigEntityName.Default))
+}
+
+object QuotaTypes {
+  val NoQuotas = 0
+  val ClientIdQuotaEnabled = 1
+  val UserQuotaEnabled = 2
+  val UserClientIdQuotaEnabled = 4
 }
 
+object QuotaId {
+
+  /**
+   * Sanitizes user principal to a safe value for use in MetricName
+   * and as Zookeeper node name
+   */
+  def sanitize(user: String): String = {
+    val encoded = URLEncoder.encode(user, StandardCharsets.UTF_8.name)
+    val builder = new StringBuilder
+    for (i <- 0 until encoded.length) {
+      encoded.charAt(i) match {
+        case '*' => builder.append("%2A") // Metric ObjectName treats * as pattern
+        case '+' => builder.append("%20") // Space URL-encoded as +, replace with percent encoding
+        case '_' => builder.append("%5F") // Encode underscore and replace % with underscore since some metrics tools might not handle %
```

Comment:
I was thinking that if there are tools that dont like %, it may be useful to replace % as well, and I picked underscore as the character to replace %. But having looked into it a bit more, I think this is unnecessary. % does appear elsewhere in metrics, so it should be safe to leave it. Have removed the encoding of _.